File: PEEC_Script/inspect_cache.m
% inspect_cache.m
% Diagnostic script to inspect cached wire data structure
% This will help identify why foil wire dimensions aren't updating

fprintf('\n=== WIRE CACHE INSPECTION ===\n\n');

% Load the cache file
cache_file = 'openmagnetics_cache.mat';
if ~exist(cache_file, 'file')
    fprintf('ERROR: Cache file not found: %s\n', cache_file);
    return;
end

fprintf('Loading cache file: %s\n', cache_file);
cache_data = load(cache_file);

% Check what's in the cache
if ~isfield(cache_data, 'wire_database')
    fprintf('ERROR: No wire_database field in cache\n');
    return;
end

wire_db = cache_data.wire_database;
wire_names = fieldnames(wire_db);
fprintf('Total wires in cache: %d\n\n', length(wire_names));

% Find foil wires (look for wires with 'Foil' in the name)
fprintf('=== FOIL WIRES ===\n');
foil_wires = {};
for i = 1:length(wire_names)
    name = wire_names{i};
    if contains(name, 'Foil') || contains(name, 'foil')
        foil_wires{end+1} = name;
    end
end

fprintf('Found %d foil wires:\n', length(foil_wires));
for i = 1:length(foil_wires)
    fprintf('  %d. %s\n', i, foil_wires{i});
end
fprintf('\n');

% Inspect the first few foil wires in detail
fprintf('=== DETAILED INSPECTION OF FOIL WIRES ===\n\n');
num_to_inspect = min(5, length(foil_wires));

for i = 1:num_to_inspect
    wire_name = foil_wires{i};
    wire = wire_db.(wire_name);

    fprintf('--- Wire %d: %s ---\n', i, wire_name);
    fprintf('Fields present: ');
    fields = fieldnames(wire);
    fprintf('%s', strjoin(fields, ', '));
    fprintf('\n');

    % Check for dimensional fields
    fprintf('Dimensional data:\n');

    % Check foil-specific fields
    if isfield(wire, 'foil_width')
        fprintf('  foil_width: %.6f m (%.3f mm)\n', wire.foil_width, wire.foil_width*1e3);
    else
        fprintf('  foil_width: NOT FOUND\n');
    end

    if isfield(wire, 'foil_thickness')
        fprintf('  foil_thickness: %.6f m (%.3f mm)\n', wire.foil_thickness, wire.foil_thickness*1e3);
    else
        fprintf('  foil_thickness: NOT FOUND\n');
    end

    % Check rect fields
    if isfield(wire, 'rect_width')
        fprintf('  rect_width: %.6f m (%.3f mm)\n', wire.rect_width, wire.rect_width*1e3);
    else
        fprintf('  rect_width: NOT FOUND\n');
    end

    if isfield(wire, 'rect_height')
        fprintf('  rect_height: %.6f m (%.3f mm)\n', wire.rect_height, wire.rect_height*1e3);
    else
        fprintf('  rect_height: NOT FOUND\n');
    end

    % Check generic fields
    if isfield(wire, 'width')
        fprintf('  width: %.6f m (%.3f mm)\n', wire.width, wire.width*1e3);
    else
        fprintf('  width: NOT FOUND\n');
    end

    if isfield(wire, 'thickness')
        fprintf('  thickness: %.6f m (%.3f mm)\n', wire.thickness, wire.thickness*1e3);
    else
        fprintf('  thickness: NOT FOUND\n');
    end

    % Check conductor shape
    if isfield(wire, 'conductor_shape')
        fprintf('  conductor_shape: %s\n', wire.conductor_shape);
    else
        fprintf('  conductor_shape: NOT FOUND\n');
    end

    % Check area
    if isfield(wire, 'area')
        fprintf('  area: %.6e m^2\n', wire.area);
    else
        fprintf('  area: NOT FOUND\n');
    end

    fprintf('\n');
end

% Look for the specific wires mentioned by the user
fprintf('=== CHECKING SPECIFIC WIRES MENTIONED ===\n');
test_wires = {'Foil_0_038', 'Foil_0_005'};

for i = 1:length(test_wires)
    wire_name = test_wires{i};
    % Try to find this wire (field names are sanitized)
    wire_name_sanitized = strrep(wire_name, '.', '_');

    if isfield(wire_db, wire_name_sanitized)
        fprintf('\nWire: %s (field: %s) - FOUND\n', wire_name, wire_name_sanitized);
        wire = wire_db.(wire_name_sanitized);

        fprintf('  All fields: %s\n', strjoin(fieldnames(wire), ', '));

        if isfield(wire, 'foil_width') && isfield(wire, 'foil_thickness')
            fprintf('  Dimensions: %.3f mm x %.3f mm\n', ...
                wire.foil_width*1e3, wire.foil_thickness*1e3);
        elseif isfield(wire, 'rect_width') && isfield(wire, 'rect_height')
            fprintf('  Dimensions: %.3f mm x %.3f mm\n', ...
                wire.rect_width*1e3, wire.rect_height*1e3);
        else
            fprintf('  WARNING: No dimensional data found!\n');
        end
    else
        fprintf('\nWire: %s (tried field: %s) - NOT FOUND\n', wire_name, wire_name_sanitized);
    end
end

fprintf('\n=== INSPECTION COMPLETE ===\n');

File: PEEC_Script/inspect_cache.py
#!/usr/bin/env python3
"""
Inspect the openmagnetics_cache.mat file to diagnose wire dimension issues
"""

import scipy.io
import numpy as np
import sys

def print_struct_fields(struct, indent=0):
    """Recursively print struct fields"""
    prefix = "  " * indent
    if isinstance(struct, dict):
        for key, value in struct.items():
            if isinstance(value, (np.ndarray, list)) and len(str(value)) > 100:
                print(f"{prefix}{key}: <array/data>")
            elif isinstance(value, dict):
                print(f"{prefix}{key}:")
                print_struct_fields(value, indent+1)
            else:
                print(f"{prefix}{key}: {value}")

def inspect_cache():
    print("\n=== WIRE CACHE INSPECTION ===\n")

    # Load the cache file
    cache_file = 'openmagnetics_cache.mat'
    try:
        print(f"Loading cache file: {cache_file}")
        cache_data = scipy.io.loadmat(cache_file, struct_as_record=False, squeeze_me=True)
        print("Cache loaded successfully\n")
    except FileNotFoundError:
        print(f"ERROR: Cache file not found: {cache_file}")
        return
    except Exception as e:
        print(f"ERROR loading cache: {e}")
        return

    # Check for wire_database
    if 'wire_database' not in cache_data:
        print("ERROR: No wire_database field in cache")
        print(f"Available fields: {list(cache_data.keys())}")
        return

    wire_db = cache_data['wire_database']
    print(f"Wire database type: {type(wire_db)}")

    # Try to access wire data
    if hasattr(wire_db, '_fieldnames'):
        wire_names = wire_db._fieldnames
        print(f"Total wires in cache: {len(wire_names)}\n")

        # Find foil wires
        print("=== FOIL WIRES ===")
        foil_wires = [name for name in wire_names if 'foil' in name.lower()]
        print(f"Found {len(foil_wires)} foil wires:")
        for i, name in enumerate(foil_wires[:20], 1):  # Show first 20
            print(f"  {i}. {name}")
        print()

        # Inspect first few foil wires in detail
        print("=== DETAILED INSPECTION OF FOIL WIRES ===\n")
        for i, wire_name in enumerate(foil_wires[:5], 1):
            print(f"--- Wire {i}: {wire_name} ---")
            wire = getattr(wire_db, wire_name)

            if hasattr(wire, '_fieldnames'):
                fields = wire._fieldnames
                print(f"Fields present: {', '.join(fields)}")

                # Check dimensional fields
                print("Dimensional data:")
                dim_fields = ['foil_width', 'foil_thickness', 'rect_width', 'rect_height',
                              'width', 'thickness', 'conductor_shape', 'area']

                for field in dim_fields:
                    if field in fields:
                        value = getattr(wire, field)
                        if field in ['foil_width', 'foil_thickness', 'rect_width',
                                     'rect_height', 'width', 'thickness']:
                            if isinstance(value, (int, float, np.number)):
                                print(f"  {field}: {value:.6f} m ({value*1e3:.3f} mm)")
                            else:
                                print(f"  {field}: {value}")
                        else:
                            print(f"  {field}: {value}")
                    else:
                        print(f"  {field}: NOT FOUND")
            else:
                print(f"Cannot inspect wire structure: {type(wire)}")
            print()

        # Check specific wires mentioned by user
        print("=== CHECKING SPECIFIC WIRES MENTIONED ===")
        test_wires = ['Foil_0_038', 'Foil_0_005']

        for wire_name in test_wires:
            # Try both with and without sanitization
            variants = [wire_name, wire_name.replace('.', '_')]

            found = False
            for variant in variants:
                if hasattr(wire_db, variant):
                    found = True
                    print(f"\nWire: {wire_name} (field: {variant}) - FOUND")
                    wire = getattr(wire_db, variant)

                    if hasattr(wire, '_fieldnames'):
                        fields = wire._fieldnames
                        print(f"  All fields: {', '.join(fields)}")

                        # Check for dimensional data
                        has_foil = hasattr(wire, 'foil_width') and hasattr(wire, 'foil_thickness')
                        has_rect = hasattr(wire, 'rect_width') and hasattr(wire, 'rect_height')

                        if has_foil:
                            w = getattr(wire, 'foil_width')
                            t = getattr(wire, 'foil_thickness')
                            print(f"  Dimensions (foil): {w*1e3:.3f} mm x {t*1e3:.3f} mm")
                        elif has_rect:
                            w = getattr(wire, 'rect_width')
                            h = getattr(wire, 'rect_height')
                            print(f"  Dimensions (rect): {w*1e3:.3f} mm x {h*1e3:.3f} mm")
                        else:
                            print(f"  WARNING: No dimensional data found!")
                    break

            if not found:
                print(f"\nWire: {wire_name} - NOT FOUND in cache")

    else:
        print(f"Cannot access wire database structure: {type(wire_db)}")
        print(f"Available attributes: {dir(wire_db)}")

    print("\n=== INSPECTION COMPLETE ===")

if __name__ == '__main__':
    inspect_cache()
